mlflow/store/sqlalchemy_store.py

```diff
@@ -2,27 +2,92 @@
     INVALID_STATE, RESOURCE_DOES_NOT_EXIST
+from mlflow.utils.file_utils import build_path, exists, mkdir
 from mlflow.utils.mlflow_tags import MLFLOW_PARENT_RUN_ID, MLFLOW_RUN_NAME
 from mlflow.utils.search_utils import does_run_match_clause
 
 
+DEFAULT_ARTIFACT_ROOT = "artifact_root"
+
+
+class DbHelper:
+    def __init__(self, db_type, session):
+        if db_type not in DB_ENGINES:
+            raise MlflowException("Unsupported DB engine '{}'. "
+                                  "Valid types are: {}'".format(db_type, DB_ENGINES))
+        self.db_type = db_type
+        self.session = session
+
+    def __enter__(self):
+        if self.db_type == MYSQL:
+            self.session.execute("SELECT @@GLOBAL.sql_mode;")
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.db_type == MYSQL:
+            self.session.execute("SELECT @@GLOBAL.sql_mode='';")
+
+
 class SqlAlchemyStore(AbstractStore):
+    ARTIFACTS_FOLDER_NAME = "artifacts"
+
+    def __init__(self, db_uri, default_artifact_root=None):
 
-    def __init__(self, db_uri):
         super(SqlAlchemyStore, self).__init__()
+        self.db_uri = db_uri
+        self.artifact_root_uri = default_artifact_root or DEFAULT_ARTIFACT_ROOT
         self.engine = sqlalchemy.create_engine(db_uri)
         Base.metadata.create_all(self.engine)
         Base.metadata.bind = self.engine
         db_session = sqlalchemy.orm.sessionmaker(bind=self.engine)
         self.session = db_session()
 
+        if self.artifact_root_uri == DEFAULT_ARTIFACT_ROOT and not exists(DEFAULT_ARTIFACT_ROOT):
+            mkdir(DEFAULT_ARTIFACT_ROOT)
+
+        if len(self.list_experiments()) == 0:
+            self._create_default_experiment(db_uri.split(':')[0])
+
+    def _create_default_experiment(self, db_type):
+        """
+        MLflow UI and client code expects a default experiment with ID 0.
+        This method uses SQL insert statement to create the default experiment as a hack, since
+        experiment table uses 'experiment_id' column is a PK and is also set to auto increment.
+        MySQL and other implementation do not allow value '0' for such cases.
+
+        ToDo: Identify a less hack mechanism to create default experiment 0
+        """
+        table = SqlExperiment.__tablename__
+        default_experiment = {
+            SqlExperiment.experiment_id.name: 0,
+            SqlExperiment.name.name: 'Default',
+            SqlExperiment.artifact_location.name: self._get_artifact_location(0),
+            SqlExperiment.lifecycle_stage.name: LifecycleStage.ACTIVE
+        }
+
+        def decorate(s):
+            if isinstance(s, str):
+                return "'{}'".format(s)
+            else:
+                return "{}".format(s)
+
+        # Get a list of keys to ensure we have a deterministic ordering
+        columns = list(default_experiment.keys())
+        values = ", ".join([decorate(default_experiment.get(c)) for c in columns])
```

Comment:
There must be some way to SQL inject here by setting a clever artifact URI...

Reply:
Let's fix this separately...